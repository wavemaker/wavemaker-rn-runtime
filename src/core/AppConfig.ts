export interface Drawer {
    setContent: (c: React.ReactNode) => void;
    getContent: () => React.ReactNode;
    setAnimation: (animation: string) => void;
    getAnimation: () => string;
}

export default interface AppConfig {
    assets: any;
    appProperties: any;
    appLocale: any;
    url: string;
    loadApp: boolean;
    refresh: (complete?: boolean) => void,
    currentPage?: any;
    pages?: any[];
    landingPage: string;
    partials?: any[];
    drawer: Drawer;
    app: any;
    setDrawerContent: any;
    theme: any;
    drawerType: any;
    getServiceDefinitions: any;
    loggedInUser: any;
    selectedLocale: string;
}