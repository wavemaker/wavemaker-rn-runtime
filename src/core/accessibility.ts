import { AccessibilityInfo } from 'react-native';
import { isAndroid, removeUndefinedKeys } from './utils';

let _isScreenReaderEnabled = false;

AccessibilityInfo.addEventListener(
  'screenReaderChanged',
  flag => {
    _isScreenReaderEnabled = flag;
  },
);

export const isScreenReaderEnabled = () => _isScreenReaderEnabled;

async function getScreenReaderStatus() {
  _isScreenReaderEnabled = await AccessibilityInfo.isScreenReaderEnabled();
}

getScreenReaderStatus();

export enum AccessibilityWidgetType {
  BUTTON = 'button',
  PICTURE = 'picture',
  TEXT = 'text',
  NUMBER = 'number',
  TEXTAREA = 'textarea',
  SELECT = 'select',
  CHIPS = 'chips',
  CURRENCY = 'currency',
  RADIOSET = 'radioset',
  CHECKBOX = 'checkbox',
  TOGGLE = 'toggle',
  SWITCH = 'switch',
  DATE = 'date',
  VIDEO = 'video',
  PROGRESSBAR = 'progressbar',
  PROGRESSCIRCLE = 'progresscircle',
  LABEL = 'label',
  ANCHOR = 'anchor',
  MESSAGE = 'message',
  SEARCH = 'search',
  ICON = 'icon',
  NAV = 'nav',
  POVOVER = 'popover',
  WEBVIEW = 'webview',
  LINECHART = 'linechart',
};

  
export type AccessibilityPropsType = {
  accessible?: boolean;
  accessibilityLabel?: string;
  accessibilityLabelledBy?: string;
  accessibilityHint?: string;
  accessibilityRole?: 'button' | 'link' | 'header' | 'search' | 'image' | 'imagebutton' | 'none' | 'summary' | 'text' | 'progressbar' | 'grid' | 'alert';
  accessibilityState?: {
    disabled?: boolean;
    selected?: boolean;
    checked?: boolean;
    expanded?: boolean;
  };
  accessibilityValue?: {
    min?: number;
    max?: number;
    now?: number;
    text?: string;
  };
  accessibilityActions?: Array<{
    name: string;
    label?: string;
  }>;
  accessibilityLiveRegion?: 'none' | 'polite' | 'assertive';
  accessibilityLanguage?: any;
  accessibilityElementsHidden?: boolean;
  accessibilityViewIsModal?: boolean;
};
  

export const getAccessibilityProps = (widgetType: AccessibilityWidgetType, accessibilityProps: AccessibilityPropsType | any) => {
  let props: AccessibilityPropsType = {accessible: true};
  if (!_isScreenReaderEnabled) {
    return {};
  }
  switch (widgetType) {
    case AccessibilityWidgetType.BUTTON:
    case AccessibilityWidgetType.TEXT:
    case AccessibilityWidgetType.NUMBER:
    case AccessibilityWidgetType.TEXTAREA:
    case AccessibilityWidgetType.SELECT:
    case AccessibilityWidgetType.CURRENCY:
    case AccessibilityWidgetType.TOGGLE:
    case AccessibilityWidgetType.DATE:
    case AccessibilityWidgetType.LABEL:
    case AccessibilityWidgetType.ANCHOR:
    case AccessibilityWidgetType.MESSAGE:    
    case AccessibilityWidgetType.SEARCH: 
    case AccessibilityWidgetType.PICTURE: 
    case AccessibilityWidgetType.ICON:
    case AccessibilityWidgetType.NAV:
    case AccessibilityWidgetType.POVOVER:
    case AccessibilityWidgetType.WEBVIEW:
    case AccessibilityWidgetType.LINECHART:  
    case AccessibilityWidgetType.VIDEO: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityHint = accessibilityProps.hint;
      props.accessibilityRole = accessibilityProps.accessibilityrole;

      if (
        widgetType === AccessibilityWidgetType.BUTTON ||
        widgetType === AccessibilityWidgetType.TEXT ||
        widgetType === AccessibilityWidgetType.NUMBER ||
        widgetType === AccessibilityWidgetType.TEXTAREA ||
        widgetType === AccessibilityWidgetType.SELECT ||
        widgetType === AccessibilityWidgetType.TOGGLE ||
        widgetType === AccessibilityWidgetType.DATE
      ) {
        props.accessibilityState = { disabled: accessibilityProps.disabled };
      }
      if (
        (widgetType === AccessibilityWidgetType.TEXT ||
          widgetType === AccessibilityWidgetType.NUMBER ||
          widgetType === AccessibilityWidgetType.TEXTAREA ||
          widgetType === AccessibilityWidgetType.SELECT ||
          widgetType === AccessibilityWidgetType.CURRENCY ||
          widgetType === AccessibilityWidgetType.TOGGLE) &&
        isAndroid()
      ) {
        props.accessibilityLabelledBy =
          accessibilityProps.accessibilitylabelledby;
      }
      if (
        widgetType === AccessibilityWidgetType.NUMBER ||
        widgetType === AccessibilityWidgetType.CURRENCY
      ) {
        props.accessibilityValue = {
          min: accessibilityProps.minvalue,
          max: accessibilityProps.maxvalue,
        };
      }
      if (widgetType === AccessibilityWidgetType.SELECT) {
        props.accessibilityState = {
          ...props.accessibilityState,
          expanded: accessibilityProps.expanded,
        };
      }
      if (widgetType === AccessibilityWidgetType.TOGGLE) {
        props.accessibilityState = {
          ...props.accessibilityState,
          selected: accessibilityProps.selected,
        };
      }
      break;
    }

    case AccessibilityWidgetType.CHIPS: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityHint = accessibilityProps.hint;
      props.accessibilityState = {
        disabled: accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }

    case AccessibilityWidgetType.RADIOSET: {
      props.accessibilityState = {
        disabled: accessibilityProps.readonly || accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }

    case AccessibilityWidgetType.CHECKBOX: {
      props.accessibilityState = {
        disabled: accessibilityProps.readonly ||  accessibilityProps.disabled,
        checked: accessibilityProps.checked,
      };
      break;
    }

    case AccessibilityWidgetType.SWITCH: {
      props.accessibilityState = {
        disabled: accessibilityProps.disabled,
        selected: accessibilityProps.selected,
      };
      break;
    }
    case AccessibilityWidgetType.PROGRESSBAR:
    case AccessibilityWidgetType.PROGRESSCIRCLE: {
      props.accessibilityLabel = accessibilityProps.accessibilitylabel || accessibilityProps.caption;
      props.accessibilityRole = accessibilityProps.accessibilityrole;
      break;
    }
    default:
      break;
  }

  const finalProps = removeUndefinedKeys(props);
  // console.log('finalProps', finalProps)

  return finalProps;
}
  