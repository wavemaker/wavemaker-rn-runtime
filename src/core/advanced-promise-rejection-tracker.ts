/**
 * Advanced Promise Rejection Tracker
 *
 * Provides comprehensive promise rejection tracking across different JavaScript engines.
 * This module addresses the critical gap where API call failures and async function errors
 * are not caught by React Error Boundaries, ensuring all promise rejections are handled.
 *
 * Engine Support:
 * - Web: Uses browser's 'unhandledrejection' event
 * - Hermes: Uses HermesInternal.enablePromiseRejectionTracker
 * - JSC: Uses React Native's 'promise/setimmediate/rejection-tracking' module
 *
 * Coverage Includes:
 * - API call failures (fetch, axios, any HTTP library)
 * - Async function errors (unhandled throws in async/await)
 * - Manual Promise.reject() calls
 * - Third-party library promise rejections
 *
 * Research Notes:
 * This implementation was developed through extensive research into React Native's
 * promise handling across different engines. Each engine has its own optimal approach:
 * - Hermes provides a native tracking API that hooks into the engine's internals
 * - JSC can use React Native's internal promise tracking module
 * - Web browsers have standardized unhandledrejection events
 *
 * The approach automatically detects the best available method and falls back
 * gracefully when native tracking is unavailable.
 */

import { Platform } from 'react-native';

// TypeScript declarations for global objects
declare global {
  interface Global {
    HermesInternal?: {
      enablePromiseRejectionTracker?: (options: {
        allRejections: boolean;
        onUnhandled: (id: number, error: any) => void;
        onHandled: (id: number) => void;
      }) => void;
    };
  }
}

export interface PromiseRejectionHandler {
  (error: any): void;
}

/**
 * Multi-engine promise rejection tracking implementation
 */
export class AdvancedPromiseRejectionTracker {
  private onRejectionHandler: PromiseRejectionHandler;
  private cleanupFunctions: (() => void)[] = [];

  constructor(onRejection: PromiseRejectionHandler) {
    this.onRejectionHandler = onRejection;
  }

  /**
   * Setup promise rejection tracking based on available APIs
   * Returns true if native tracking was enabled, false if fallback used
   */
  public setup(): boolean {
    // Automatically detect and setup the best available promise rejection tracking method

    if (Platform.OS === 'web') {
      return this.setupWebTracking();
    }

    // For mobile platforms, try native tracking methods in order of preference
    if (this.setupHermesTracking()) {
      return true;
    }

    // Fallback to React Native internal tracking for JSC
    if (this.setupReactNativeTracking()) {
      return true;
    }

    // No native tracking available - this should rarely happen in modern RN versions
    return false;
  }

  /**
   * Method 1: Web Environment - Use unhandledrejection event
   */
  private setupWebTracking(): boolean {
    try {
      if (typeof window === 'undefined' || !window.addEventListener) {
        return false;
      }

      const handler = (event: PromiseRejectionEvent) => {
        event.preventDefault(); // Prevent browser's default handling
        const error =
          event.reason instanceof Error
            ? event.reason
            : new Error(String(event.reason));
        this.onRejectionHandler(error);
      };

      window.addEventListener('unhandledrejection', handler);
      this.cleanupFunctions.push(() => {
        window.removeEventListener('unhandledrejection', handler);
      });

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 2: Hermes Engine - Use HermesInternal API
   */
  private setupHermesTracking(): boolean {
    try {
      // Check if Hermes engine is available
      const hermesInternal = (global as any).HermesInternal;
      if (
        hermesInternal &&
        typeof hermesInternal.enablePromiseRejectionTracker === 'function'
      ) {
        hermesInternal.enablePromiseRejectionTracker({
          allRejections: true, // Track both handled and unhandled rejections
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - could be used for analytics in the future
          },
        });

        // Cleanup function (if available)
        this.cleanupFunctions.push(() => {
          // Hermes doesn't provide a disable method, but we can set a no-op handler
          if (hermesInternal?.enablePromiseRejectionTracker) {
            hermesInternal.enablePromiseRejectionTracker({
              allRejections: false,
              onUnhandled: () => {},
              onHandled: () => {},
            });
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 3: React Native Internal - Use promise/setimmediate/rejection-tracking for JSC
   */
  private setupReactNativeTracking(): boolean {
    try {
      // Try to access React Native's internal promise tracking module
      const tracking = require('promise/setimmediate/rejection-tracking');

      if (tracking && typeof tracking.enable === 'function') {
        tracking.enable({
          allRejections: true,
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - tracking for potential analytics
          },
        });

        this.cleanupFunctions.push(() => {
          // Try to disable tracking if possible
          if (tracking.disable && typeof tracking.disable === 'function') {
            tracking.disable();
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Clean up all tracking mechanisms
   */
  public cleanup(): void {
    // Clean up all tracking mechanisms and restore original state
    this.cleanupFunctions.forEach((cleanup) => {
      try {
        cleanup();
      } catch (error) {
        // Silently handle cleanup errors to avoid disrupting app shutdown
      }
    });
    this.cleanupFunctions = [];
  }
}

/**
 * Helper function to detect JavaScript engine (simplified for Web/Hermes/JSC)
 */
export function detectJavaScriptEngine(): 'web' | 'hermes' | 'jsc' {
  if (Platform.OS === 'web') {
    return 'web';
  }

  if (typeof (global as any).HermesInternal !== 'undefined') {
    return 'hermes';
  }

  // For non-Hermes engines on mobile, assume JSC (most common)
  return 'jsc';
}

/**
 * Public API for easy integration
 */
/**
 * Sets up advanced promise rejection tracking with automatic engine detection
 * @param onRejection Callback function to handle unhandled promise rejections
 * @returns AdvancedPromiseRejectionTracker instance for cleanup and testing
 */
export function setupAdvancedPromiseRejectionTracking(
  onRejection: PromiseRejectionHandler
): AdvancedPromiseRejectionTracker {
  const tracker = new AdvancedPromiseRejectionTracker(onRejection);
  tracker.setup();
  return tracker;
}
