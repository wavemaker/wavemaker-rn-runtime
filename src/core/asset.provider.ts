import React from 'react';

const AssetContext = React.createContext<(path: string) => number | string>(null as any);

export const AssetProvider = AssetContext.Provider;
export const AssetConsumer = AssetContext.Consumer;