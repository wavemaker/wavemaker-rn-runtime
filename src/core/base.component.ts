import { merge } from 'lodash';
import React, { ReactNode } from 'react';
import BASE_THEME, { DEFAULT_CLASS, DEFAULT_STYLE } from '../styles/theme';
import { PropsProvider } from './props.provider';

export interface ComponentState<T extends BaseProps> {
    props: T;
}

export class BaseProps {
    name: string = null as any;
    themeToUse = BASE_THEME;
    show = true;
    styles: any = null;
    onInit: Function = null as any;
    onDestroy: Function = null as  any;
}

export abstract class BaseComponent<T extends BaseProps> extends React.Component<T, ComponentState<T>> {
    public styles: any = DEFAULT_STYLE;
    private propertyProvider: PropsProvider<T>;
    public proxy: BaseComponent<T>;

    constructor(markupProps: T, public defaultClass = DEFAULT_CLASS, private defaultStyles = DEFAULT_STYLE, defaultProps?: T) {
        super(markupProps);
        this.propertyProvider = new PropsProvider<T>(
            Object.assign(defaultProps || {}, markupProps), 
            (name: string, $new: any, $old: any) => {
                this.onPropertyChange(name, $new, $old);
            });
        this.state = {
            props: this.propertyProvider.get()
        };
        //@ts-ignore
        this.proxy = (new Proxy(this, {
            get: (target, prop, receiver): any => {
                const propName = prop.toString();
                if (this.propertyProvider.has(propName)) {
                    //@ts-ignore
                    return this.state.props[propName];
                }
                return Reflect.get(target, prop, receiver);
            },
            set: (target: any, prop, value: any): any => {
                const propName = prop.toString();
                if (this.propertyProvider.has(propName)) {
                    // @ts-ignore
                    this.state.props[propName] = value;
                    this.setState({
                        props: this.propertyProvider.get()
                    })
                    return true;
                } else {
                    return Reflect.set(target, prop, value);
                }
            }
        }));
    }

    onPropertyChange(name: string, $new: any, $old: any) {

    }

    shouldComponentUpdate(nextProps: T, nextState: ComponentState<T>, nextContext: any) {
        if (this.propertyProvider.check(nextProps)) {
            this.setState({
                props: this.propertyProvider.get()
            });
            return true;
        }
        return !!(super.shouldComponentUpdate && super.shouldComponentUpdate(nextProps, nextState, nextContext));
    }

    componentDidMount() {
        this.props.onInit && this.props.onInit(this.proxy);
    }

    componentWillUnmount() {
        this.props.onDestroy && this.props.onDestroy(this);
    }

    invokeEventCallback(eventName: string, args: any[]) {
        //@ts-ignore
        const callBack: Function = this.props[eventName];
        if (callBack) {
            callBack.apply(this.proxy, args);
        }
        if (eventName === 'onTap') {
            this.invokeEventCallback('onClick', args);
        }
    }

    public render(): ReactNode {
        this.styles = merge({}, this.props.themeToUse.getStyle(this.defaultClass) || this.defaultStyles, this.props.styles);
        return null;
    }
}