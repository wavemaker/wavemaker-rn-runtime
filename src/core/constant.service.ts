export class ConstantService {
   VALIDATOR = {
    REQUIRED: 'required',
    MAXCHARS: 'maxchars',
    MINVALUE: 'minvalue',
    MAXVALUE: 'maxvalue',
    REGEXP: 'regexp',
    MINDATE: 'mindate',
    MAXDATE: 'maxdate',
    MINTIME: 'mintime',
    MAXTIME: 'maxtime',
    EXCLUDEDATES: 'excludedates',
    EXCLUDEDAYS: 'excludedays'
  };
}

export default new ConstantService();
