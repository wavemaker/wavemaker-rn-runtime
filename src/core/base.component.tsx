import { isEqual, merge } from 'lodash';
import React, { ReactNode } from 'react';
import { TextStyle } from 'react-native';
import { ROOT_LOGGER } from '@wavemaker/app-rn-runtime/core/logger'
import BASE_THEME, { DEFAULT_CLASS, DEFAULT_STYLE, NamedStyles, AllStyle, ThemeConsumer } from '../styles/theme';
import { PropsProvider } from './props.provider';

export const WIDGET_LOGGER = ROOT_LOGGER.extend('widget');

export class BaseComponentState<T extends BaseProps> {
    public props = {} as T;
}

export type BaseStyles = NamedStyles<any> & {
    root: AllStyle,
    text: TextStyle
}

export interface LifecycleListener {
    onComponentInit: (c: BaseComponent<any, any, any>) => void;
    onComponentDestroy: (c: BaseComponent<any, any, any>) => void;
}

export class BaseProps {
    id?: string = null as any;
    name?: string = null as any;
    key?: string = null as any;
    show? = true;
    styles?: any = null;
    listener?: LifecycleListener = null as any;
}

export abstract class BaseComponent<T extends BaseProps, S extends BaseComponentState<T>, L extends BaseStyles> extends React.Component<T, S> {
    public styles: L = {
        root: {},
        text: {}
    } as L;
    private propertyProvider: PropsProvider<T>;
    public proxy: BaseComponent<T, S, L>;
    private initialized = false;
    public cleanup = [] as Function[];
    public theme = BASE_THEME;

    constructor(markupProps: T, public defaultClass = DEFAULT_CLASS, private defaultStyles?: L, defaultProps?: T, defaultState?: S) {
        super(markupProps);
        this.state = (defaultState || {} as S);
        this.propertyProvider = new PropsProvider<T>(
            Object.assign(defaultProps || {}, markupProps),
            (name: string, $new: any, $old: any) => {
                WIDGET_LOGGER.debug(() => `${this.props.name ?? this.constructor.name}: ${name} changed from ${$old} to ${$new}`);
                this.onPropertyChange(name, $new, $old);
                if (name === 'styles') {
                    this.styles = null as any;
                }
            });
        //@ts-ignore
        this.state.props =this.propertyProvider.get();
        this.propertyProvider.check();
        //@ts-ignore
        this.proxy = (new Proxy(this, {
            get: (target, prop, receiver): any => {
                const propName = prop.toString();
                if (this.propertyProvider.has(propName)) {
                    //@ts-ignore
                    return this.state.props[propName];
                }
                return Reflect.get(target, prop, receiver);
            },
            set: (target: any, prop, value: any): any => {
                const propName = prop.toString();
                if (this.propertyProvider.has(propName)) {
                    // @ts-ignore
                    this.state.props[propName] = value;
                    this.setState(() => ({
                        props: this.propertyProvider.get()
                    }))
                    return true;
                } else {
                    return Reflect.set(target, prop, value);
                }
            }
        }));
    }

    onPropertyChange(name: string, $new: any, $old: any) {

    }

    updateState(state: S) {
      if (state.props) {
        Object.keys(state.props).forEach((k) => {
          //@ts-ignore
          this.state.props[k] = state.props[k];
        });
        state.props = this.state.props
      }
        if (!this.initialized) {
            Object.keys(state).forEach((key) => {
                //@ts-ignore
                this.state[key] = state[key];
            });
        } else {
            this.setState(() => state);
        }
    }

    shouldComponentUpdate(nextProps: T, nextState: S, nextContext: any) {
        return this.propertyProvider.check(nextProps) || !isEqual(nextState, this.state);
    }

    componentDidMount() {
        if (this.props.listener) {
            this.props.listener.onComponentInit(this.proxy);
        }
        this.initialized = true;
    }

    componentWillUnmount() {
        if (this.props.listener) {
            this.props.listener.onComponentDestroy(this.proxy);
        }
        this.cleanup.forEach(f => f && f());
    }

    invokeEventCallback(eventName: string, args: any[]) {
        //@ts-ignore
        const callBack: Function = this.props[eventName];
        args = args && args.map(a => (a === this) ? this.proxy : a)
        if (callBack) {
            callBack.apply(this.proxy, args);
        }
        if (eventName === 'onTap') {
          this.invokeEventCallback && this.invokeEventCallback('onClick', args);
        }
    }

    protected abstract renderWidget(props: T): ReactNode;

    public render(): ReactNode {
        WIDGET_LOGGER.info(() => `${this.props.name ?? this.constructor.name} is rendering.`);
        const props = this.state.props;
        return props.show !== false ?
            (<ThemeConsumer>{(theme) => {
                this.theme = theme || BASE_THEME;
                this.styles = this.styles || merge({}, this.theme.getStyle(this.defaultClass) || this.defaultStyles, this.props.styles);
                return this.renderWidget(this.state.props);
            }}</ThemeConsumer>) : null;
    }
}
