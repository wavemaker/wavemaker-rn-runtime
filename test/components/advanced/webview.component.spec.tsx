import React, { createRef } from 'react';
import { AccessibilityInfo, Platform, View } from 'react-native';
import { act, fireEvent, render, waitFor } from '@testing-library/react-native';
import WmWebview from '@wavemaker/app-rn-runtime/components/advanced/webview/webview.component';
import * as accessibilityUtils from '@wavemaker/app-rn-runtime/core/accessibility';

describe('Test Webview component', () => {
  const baseProps = {
    id: 'wm',
    name: 'webview-component',
    src: 'https://example.com',
    incognito: false,
    onLoad: jest.fn(),
  };

  const invokeEventCallbackMock = jest.spyOn(
    WmWebview.prototype,
    'invokeEventCallback'
  );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render WmWebview component with given src', async () => {
    const tree = render(<WmWebview {...baseProps} />);
    const webview = tree.getByTestId('wm_web_view');

    // await waitFor(()=>{
    //   expect(webview.props.source).toBeDefined();
    // })
    expect(tree).toMatchSnapshot();
  });

  test('should render component with accessibility props', async () => {
    const props = {
      ...baseProps,
      accessibilitylabel: 'wm-web-view',
      hint: 'this is a web view',
      accessibilityrole: 'web-view-wm',
    };
    const tree = render(<WmWebview {...props} />);

    expect(tree.getByTestId('wm_web_view')).toBeDefined();
    expect(tree.getByLabelText('wm-web-view')).toBeDefined();
    expect(tree.getByAccessibilityHint('this is a web view')).toBeDefined();
    expect(tree.getByTestId('wm_web_view').props.accessibilityRole).toBe(
      'web-view-wm'
    );
    expect(tree).toMatchSnapshot();
  });

  test('should handle onLoad callback when content is loaded', async () => {
    const tree = render(<WmWebview {...baseProps} />);
    const webview = tree.getByTestId('wm_web_view');

    const event = {
      nativeEvent: {
        title: 'Example Page',
        url: 'https://example.com',
      },
    };

    fireEvent(webview, 'onLoadEnd', event);

    await waitFor(() => {
      expect(baseProps.onLoad).toHaveBeenCalledWith(
        event,
        expect.any(WmWebview)
      );
      expect(tree).toMatchSnapshot();
    });
  });

  test('should inject JavaScript and handle result', async () => {
    const customRef = createRef<WmWebview>();
    render(<WmWebview {...baseProps} ref={customRef} />);

    const invokeScriptFn = jest.spyOn(
      customRef.current.webview,
      'injectJavaScript'
    );

    customRef.current.executeScript(`function(){console.log("hello world")}`);
    expect(invokeScriptFn).toHaveBeenCalled();
    expect(invokeScriptFn).toHaveBeenCalledWith(
      expect.stringContaining(`function(){console.log("hello world")}`)
    );
  });

  // test('should handle BackHandler for Android', () => {
  //   Platform.OS = 'android';
  //   const customRef = createRef<WmWebview>()

  //   // console.log(WmWebview.prototype);

  //   // const backHandlerSpy = jest.spyOn(
  //   //   WmWebview.prototype,
  //   //   'handleBackButtonPress'
  //   // );
  //   render(<WmWebview {...baseProps} ref={customRef}/>);

  //   // act(()=>{
  //   //   customRef.current
  //   // })
  //   hardwareBackPress
  //   console.log(customRef.current);

  //   expect(backHandlerSpy).toHaveBeenCalled();
  // });

  test('should use incognito mode if incognito prop is true', () => {
    const props = {
      ...baseProps,
      incognito: true,
    };

    const { getByTestId } = render(<WmWebview {...props} />);
    const webview = getByTestId('wm_web_view');

    expect(webview.props.incognito).toBe(true);
  });

  test('should handle onMessage callback when a message is received', () => {
    const parseResultMock = jest.spyOn(WmWebview.prototype, 'parseResult');
    const customRef = createRef<WmWebview>();
    const { getByTestId } = render(
      <WmWebview {...baseProps} ref={customRef} />
    );
    const webview = getByTestId('wm_web_view');

    const event = {
      nativeEvent: {
        data: 'test message',
      },
    };

    fireEvent(webview, 'onMessage', event);
    expect(invokeEventCallbackMock).toHaveBeenCalledWith('onMessage', [
      event,
      expect.anything(),
    ]);

    act(() => {
      customRef.current.invokeJSCallbacks = {
        1: (arg) => arg,
      };
    });

    event.nativeEvent = {
      data: 'afterInjectJavaScript:1:function()',
    };

    fireEvent(webview, 'onMessage', event);
    expect(parseResultMock).toHaveBeenCalled();
  });

  test('should parse the result correctly from executeScript', () => {
    const parseResultMock = jest.spyOn(WmWebview.prototype, 'parseResult');
    const customRef = createRef<WmWebview>();
    const { getByTestId } = render(
      <WmWebview {...baseProps} ref={customRef} />
    );
    const webview = getByTestId('wm_web_view');

    act(() => {
      customRef.current.invokeJSCallbacks = {
        1: (arg) => arg,
      };
    });

    const event = {
      nativeEvent: {
        data: 'afterInjectJavaScript:1:{value: "wm-web-view"}',
      },
    };

    fireEvent(webview, 'onMessage', event);
    expect(parseResultMock).toHaveBeenCalled();
    expect(parseResultMock).toHaveBeenCalledWith('{value: "wm-web-view"}')
  });

  test('should handle insertCSS correctly', async () => {
    const executeScriptMock = jest.spyOn(WmWebview.prototype, 'executeScript');
    const customRef = createRef();
    render(<WmWebview {...baseProps} ref={customRef}/>);
    let result;

    act(()=>{
      result = customRef.current.insertCSS(
        'body { background-color: red; }'
      );
    })

    expect(executeScriptMock).toHaveBeenCalled();
    await waitFor(()=>{
      expect(result).toMatchObject({})
    })
  });

  // test('should render iframe when Platform is web', () => {
  //   jest.mock(iframe, ()=><View/>)
  //   Platform.OS = 'web';
  //   const tree = render(<WmWebview {...baseProps}/>);

  //   expect(tree.UNSAFE_getByType(<iframe/>)).toBeDefined();
  // })
});
