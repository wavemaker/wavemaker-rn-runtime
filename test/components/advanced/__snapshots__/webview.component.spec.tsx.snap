// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Webview component should handle onLoad callback when content is loaded 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      {
        "height": 0,
        "width": 0,
      }
    }
  />,
  <View
    style={{}}
  >
    <View
      style={
        [
          {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        accessible={true}
        cacheEnabled={true}
        fraudulentWebsiteWarningEnabled={true}
        hasOnFileDownload={false}
        hasOnOpenWindowEvent={false}
        incognito={false}
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptEnabled={true}
        messagingEnabled={true}
        messagingModuleName=""
        nestedScrollEnabled={true}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        scrollEnabled={true}
        sharedCookiesEnabled={true}
        style={
          [
            {
              "flex": 1,
              "overflow": "hidden",
            },
            {
              "backgroundColor": "#ffffff",
            },
            {
              "flex": 1,
            },
          ]
        }
        testID="wm_web_view"
        textInteractionEnabled={true}
        useSharedProcessPool={true}
      />
    </View>
  </View>,
]
`;

exports[`Test Webview component should render WmWebview component with given src 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      {
        "height": 0,
        "width": 0,
      }
    }
  />,
  <View
    style={{}}
  >
    <View
      style={
        [
          {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        accessible={true}
        cacheEnabled={true}
        fraudulentWebsiteWarningEnabled={true}
        hasOnFileDownload={false}
        hasOnOpenWindowEvent={false}
        incognito={false}
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptEnabled={true}
        messagingEnabled={true}
        messagingModuleName=""
        nestedScrollEnabled={true}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        scrollEnabled={true}
        sharedCookiesEnabled={true}
        style={
          [
            {
              "flex": 1,
              "overflow": "hidden",
            },
            {
              "backgroundColor": "#ffffff",
            },
            {
              "flex": 1,
            },
          ]
        }
        testID="wm_web_view"
        textInteractionEnabled={true}
        useSharedProcessPool={true}
      />
    </View>
  </View>,
]
`;

exports[`Test Webview component should render component with accessibility props 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      {
        "height": 0,
        "width": 0,
      }
    }
  />,
  <View
    style={{}}
  >
    <View
      style={
        [
          {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        accessibilityHint="this is a web view"
        accessibilityLabel="wm-web-view"
        accessibilityRole="web-view-wm"
        accessible={true}
        cacheEnabled={true}
        fraudulentWebsiteWarningEnabled={true}
        hasOnFileDownload={false}
        hasOnOpenWindowEvent={false}
        incognito={false}
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptEnabled={true}
        messagingEnabled={true}
        messagingModuleName=""
        nestedScrollEnabled={true}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        scrollEnabled={true}
        sharedCookiesEnabled={true}
        style={
          [
            {
              "flex": 1,
              "overflow": "hidden",
            },
            {
              "backgroundColor": "#ffffff",
            },
            {
              "flex": 1,
            },
          ]
        }
        testID="wm_web_view"
        textInteractionEnabled={true}
        useSharedProcessPool={true}
      />
    </View>
  </View>,
]
`;
